### ISSUE TITLE
Request: support for COUNT and JOIN

### ISSUE THREAD

**Issue by deakjahn** [tekartik/sqflite#395]
Very simple to do and would add convenience. These are the extensions I currently use:

    extension SQL on Database {
      Future<int> count(String table, {String where, List<dynamic> whereArgs}) async {
        final query = StringBuffer();
        query.write('SELECT COUNT(*) FROM "$table"');
        if (where != null) query.write(' WHERE $where ');
    
        String sql = query.toString();
        List<dynamic> arguments = (whereArgs != null) ? List<dynamic>.from(whereArgs) : null;
        return Sqflite.firstIntValue(await rawQuery(sql, arguments));
      }
    
      Future<List<Map<String, dynamic>>> join(String table, String join, {String using, String usingOn, bool distinct, List<String> columns, String where, List<dynamic> whereArgs, String groupBy, String having, String orderBy, int limit, int offset}) async {
        if (groupBy == null && having != null) throw ArgumentError('Having requires groupBy');
        if (using != null && usingOn != null) throw ArgumentError('Either USING or ON');
    
        final query = StringBuffer();
        query.write('SELECT ');
        if (distinct == true) query.write('DISTINCT ');
        if (columns != null && columns.isNotEmpty)
          query.writeAll(columns, ', ');
        else
          query.write('* ');
        query.write(' FROM "$table"');
        query.write(' LEFT JOIN "$join" ');
        if (using != null) query.write(' USING($using) ');
        if (usingOn != null) query.write(' ON($usingOn) ');
        if (where != null) query.write(' WHERE $where ');
        if (groupBy != null) query.write(' GROUP BY $groupBy ');
        if (having != null) query.write(' HAVING $having ');
        if (orderBy != null) query.write(' ORDER BY $orderBy ');
        if (limit != null) query.write(' LIMIT $limit ');
        if (offset != null) query.write(' OFFSET $offset ');
    
        String sql = query.toString();
        List<dynamic> arguments = (whereArgs != null) ? List<dynamic>.from(whereArgs) : null;
        return await rawQuery(sql, arguments);
      }
    }

**Comment by alextekartik** [comment]
I like the idea of using extension for this. In fact maybe sqflite should only export rawQuery, rawUpdate and rawDelete and have the helpers (query, update, count, join) in an extension (which could be on DatabaseExecutor instead of Database so that it also work in a transaction)

Unfortunately batch has a different syntax so it means duplicating join and count here (although it is less usefull here, few people use batch for queries, although it definitely has a speed gain).

It is hard to be as exhaustive on queries as people want. For example I write join in a different way. SqlBuilder is hidden on purpose as I did not feel like maintaining queries that I don't use nor test. There are (or could be) 3rd party package for writing queries. The helpers currently are based on what the Android API propose so tend to think of keeping sqflite as lean as needed.

That is said, count (which could be implemented using query(table, columns: ['COUNT(*)'] also) makes sense. But then here some people might want to have the distinct option....it never ends...

Current query helpers are not complete. I tend to write raw sql statement so before adding too many helpers, I want to be sure it matches common needs (of course count surely does).

Alternatively SqlBuilder could be exported (or could be a separate package is someone feel like starting one) to allow building any kind of queries.

**Comment by deakjahn** [comment]
Yes, count() mostly because you have to read the result in a different way and if you provide it, everybody just sees it and uses it. If you don't, they all have to Google for the solution. :-)

**Comment by dkobia** [comment]
@deakjahn @alextekartik I advanced that extension a little further to account for null safety and explicitly for `INNER joins` and `LEFT joins`. SQLite does not currently support RIGHT and FULL OUTER joins

```
enum JoinType {
  innerJoin('INNER JOIN'),
  leftJoin('LEFT JOIN');

  const JoinType(this.value);

  final String value;
}

extension SQL on Database {
  Future<int?> count(String table, {String? where, List<dynamic>? whereArgs}) async {
    final query = StringBuffer();
    query.write('SELECT COUNT(*) FROM "$table" ');
    if (where != null) query.write('WHERE $where');

    String sql = query.toString();
    List<dynamic>? arguments = (whereArgs != null) ? List<dynamic>.from(whereArgs) : null;
    return Sqflite.firstIntValue(await rawQuery(sql, arguments));
  }

  Future<List<Map<String, dynamic>>> join(
    String table,
    String join, {
    JoinType joinType = JoinType.innerJoin,
    List<String>? using,
    String? usingOn,
    bool? distinct,
    List<String>? columns,
    String? where,
    List<dynamic>? whereArgs,
    String? groupBy,
    String? having,
    String? orderBy,
    int? limit,
    int? offset,
  }) async {
    if (groupBy == null && having != null) throw ArgumentError('Having requires groupBy');
    if (using != null && usingOn != null) throw ArgumentError('Either USING or ON');

    final query = StringBuffer();
    query.write('SELECT ');
    if (distinct == true) query.write('DISTINCT ');
    if (columns != null && columns.isNotEmpty) {
      query.writeAll(columns, ', ');
    } else {
      query.write('* ');
    }
    query.write(' FROM "$table" ');
    query.write('${joinType.value} "$join" ');
    if (using != null && using.isNotEmpty) {
      query.write('USING(');
      query.writeAll(using, ',');
      query.write(') ');
    }
    if (usingOn != null) query.write('ON($usingOn) ');
    if (where != null) query.write('WHERE $where ');
    if (groupBy != null) query.write('GROUP BY $groupBy ');
    if (having != null) query.write('HAVING $having ');
    if (orderBy != null) query.write('ORDER BY $orderBy ');
    if (limit != null) query.write('LIMIT $limit ');
    if (offset != null) query.write('OFFSET $offset ');

    String sql = query.toString();
    List<dynamic>? arguments = (whereArgs != null) ? List<dynamic>.from(whereArgs) : null;
    return await rawQuery(sql, arguments);
  }
}
```
